File: scripts/narrate.py
"""
Builds the narration track from docs/narration.json.

Uses Kokoro-82M, an open-weights model that runs locally on CPU. No API key, no account,
and nothing leaves the machine. The first track used Windows SAPI ("Microsoft Zira
Desktop"), a concatenative voice that sounded like one; the second used Microsoft Edge's
neural voices, which are better but still read as a screen reader.

Each segment is synthesised at its natural rate and then padded to the length its beat
occupies in the picture. Padding rather than time-stretching keeps the speech unhurried,
and the silence lands at the end of a beat, where a viewer is reading the held line.

    python scripts/narrate.py --measure   # spoken vs beat length, writes nothing
    python scripts/narrate.py             # writes video/public/seg00.wav ... seg06.wav
"""

from __future__ import annotations

import argparse
import json
import subprocess
import sys
import warnings
from pathlib import Path

warnings.filterwarnings("ignore")

SPEC_PATH = Path("docs/narration.json")
SAMPLE_RATE = 24_000
OUTPUT_RATE = 48_000


def build(measure_only: bool, out_dir: Path) -> int:
    import numpy as np
    import soundfile as sf
    from kokoro import KPipeline

    spec = json.loads(SPEC_PATH.read_text(encoding="utf-8"))
    voice = spec["voice"]
    speed = float(spec.get("speed", 1.0))

    pipeline = KPipeline(lang_code=spec.get("langCode", "a"), repo_id="hexgrad/Kokoro-82M")
    out_dir.mkdir(parents=True, exist_ok=True)
    tmp = Path(".narration-tmp")
    tmp.mkdir(exist_ok=True)

    overruns = 0
    for i, segment in enumerate(spec["segments"]):
        chunks = [audio for _, _, audio in pipeline(segment["text"], voice=voice, speed=speed)]
        audio = np.concatenate(chunks)
        spoken = len(audio) / SAMPLE_RATE
        beat = float(segment["durationSeconds"])
        slack = beat - spoken
        if slack < 0:
            overruns += 1
        print(
            f"{segment['beat']:<12} spoken {spoken:5.1f}s / beat {beat:5.1f}s  "
            + (f"OVER by {-slack:.1f}s" if slack < 0 else f"{slack:.1f}s of room")
        )
        if measure_only:
            continue

        raw = tmp / f"seg{i:02d}.wav"
        sf.write(str(raw), audio, SAMPLE_RATE)
        # Pad to exactly the beat length. `apad` then `atrim` is exact where `-t` is not.
        subprocess.run(
            [
                "ffmpeg", "-v", "error", "-y", "-i", str(raw),
                "-af", f"apad,atrim=0:{beat},aresample={OUTPUT_RATE}",
                "-ac", "1", "-ar", str(OUTPUT_RATE),
                str(out_dir / f"seg{i:02d}.wav"),
            ],
            check=True,
        )

    if overruns:
        print(f"\n{overruns} segment(s) longer than their beat. Shorten the text.", file=sys.stderr)
        return 1
    print("\nmeasured only" if measure_only else f"\nwrote {len(spec['segments'])} files to {out_dir}")
    return 0


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--measure", action="store_true", help="print lengths, write nothing")
    parser.add_argument("--out", default="video/public", help="where the seg*.wav files go")
    args = parser.parse_args()
    sys.exit(build(args.measure, Path(args.out)))

File: scripts/narrate.mjs
/**
 * Builds the narration track from docs/narration.json.
 *
 * Uses Microsoft Edge's neural voices through `edge-tts`, which needs no API key and no
 * account. The previous track used Windows SAPI ("Microsoft Zira Desktop"), which is a
 * concatenative voice from a different decade and sounded like one.
 *
 * Each segment is rendered separately, padded to the length its beat occupies in the
 * picture, and written as seg00.wav ... seg06.wav for the Remotion composition. Padding
 * rather than stretching keeps the speech at its natural rate; the silence lands at the
 * end of a beat, where a viewer is reading the caption anyway.
 *
 *   node scripts/narrate.mjs            # writes video/public/seg*.wav
 *   node scripts/narrate.mjs --measure  # prints spoken length vs beat length, writes nothing
 */
import { readFileSync, mkdirSync } from 'node:fs'
import { execFileSync } from 'node:child_process'

const SPEC = JSON.parse(readFileSync('docs/narration.json', 'utf8'))
const OUT = process.env.OUT_DIR ?? 'video/public'
const TMP = process.env.TMP_DIR ?? '.narration-tmp'
const measureOnly = process.argv.includes('--measure')

const duration = (file) =>
  Number(execFileSync('ffprobe', ['-v', 'error', '-show_entries', 'format=duration', '-of', 'csv=p=0', file]).toString().trim())

mkdirSync(TMP, { recursive: true })
if (!measureOnly) mkdirSync(OUT, { recursive: true })

let overrun = 0
SPEC.segments.forEach((segment, i) => {
  const raw = `${TMP}/seg${String(i).padStart(2, '0')}.mp3`
  execFileSync('python', [
    '-m', 'edge_tts',
    `--voice=${SPEC.voice}`,
    // Joined with "=" on purpose: a rate like "-4%" starts with a hyphen, and argparse
    // reads a separate "-4%" as another flag rather than as this one's value.
    `--rate=${SPEC.rate}`,
    `--pitch=${SPEC.pitch}`,
    `--text=${segment.text}`,
    `--write-media=${raw}`,
  ], { stdio: 'pipe' })

  const spoken = duration(raw)
  const beat = segment.durationSeconds
  const slack = beat - spoken
  if (slack < 0) overrun += 1
  console.log(
    `${segment.beat.padEnd(12)} spoken ${spoken.toFixed(1)}s / beat ${beat.toFixed(1)}s ` +
    `${slack < 0 ? `OVER by ${(-slack).toFixed(1)}s` : `${slack.toFixed(1)}s of room`}`,
  )

  if (measureOnly) return
  // Pad to exactly the beat length. `apad` then `atrim` is exact where `-t` alone is not.
  execFileSync('ffmpeg', [
    '-v', 'error', '-y', '-i', raw,
    '-af', `apad,atrim=0:${beat},aresample=48000`,
    '-ac', '1', '-ar', '48000',
    `${OUT}/seg${String(i).padStart(2, '0')}.wav`,
  ], { stdio: 'pipe' })
})

if (overrun > 0) {
  console.error(`\n${overrun} segment(s) longer than their beat. Shorten the text or lengthen the beat.`)
  process.exit(1)
}
console.log(measureOnly ? '\nmeasured only' : `\nwrote ${SPEC.segments.length} files to ${OUT}`)
